File: web/lib/modules/ride_management/services/client_address_service.dart
import 'dart:convert';
import '../models/client_address.dart';
import '../../core/services/api_client.dart';

class ClientAddressService {
  final ApiClient _apiClient;
  final bool _ownsClient;

  ClientAddressService({ApiClient? apiClient})
    : _apiClient = apiClient ?? ApiClient(),
      _ownsClient = apiClient == null;

  Future<List<ClientAddress>> getAddresses(String clientId) async {
    final response = await _apiClient.get('/clients/$clientId/addresses');
    if (response.statusCode == 200) {
      final List<dynamic> list = jsonDecode(response.body) as List<dynamic>;
      return list
          .map((e) => ClientAddress.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  Future<ClientAddress?> saveAddress({
    required String clientId,
    required String label,
    required String address,
    double? latitude,
    double? longitude,
  }) async {
    final body = <String, dynamic>{
      'label': label,
      'address': address,
      if (latitude != null) 'latitude': latitude,
      if (longitude != null) 'longitude': longitude,
    };
    final response = await _apiClient.post(
      '/clients/$clientId/addresses',
      body,
    );
    if (response.statusCode == 201) {
      return ClientAddress.fromJson(
        jsonDecode(response.body) as Map<String, dynamic>,
      );
    }
    return null;
  }

  Future<ClientAddress?> updateAddress({
    required String clientId,
    required String addressId,
    String? label,
    List<String>? aliases,
  }) async {
    final body = <String, dynamic>{
      if (label != null) 'label': label,
      if (aliases != null) 'aliases': aliases,
    };
    final response = await _apiClient.patch(
      '/clients/$clientId/addresses/$addressId',
      body,
    );
    if (response.statusCode == 200) {
      return ClientAddress.fromJson(
        jsonDecode(response.body) as Map<String, dynamic>,
      );
    }
    return null;
  }

  Future<bool> deleteAddress(String clientId, String addressId) async {
    final response = await _apiClient.delete(
      '/clients/$clientId/addresses/$addressId',
    );
    return response.statusCode == 204;
  }

  void dispose() {
    if (_ownsClient) _apiClient.dispose();
  }
}

File: ride/src/main/scala/com/shevchyk/ride/application/service/ClientAddressService.scala
package com.shevchyk.ride.application.service

import com.shevchyk.core.domain.PersonId
import com.shevchyk.ride.domain.{ClientAddress, ClientAddressId, SaveClientAddressRequest, UpdateClientAddressRequest}
import com.shevchyk.ride.repository.ClientAddressRepository
import zio.*
import java.time.Instant

trait ClientAddressService:
  def getAddresses(clientId: PersonId): Task[List[ClientAddress]]
  def saveAddress(clientId: PersonId, req: SaveClientAddressRequest): Task[ClientAddress]

  def updateAddress(
      id: ClientAddressId,
      clientId: PersonId,
      req: UpdateClientAddressRequest
  ): Task[Option[ClientAddress]]

  def recordUsage(
      clientId: PersonId,
      address: String,
      label: String,
      lat: Option[Double],
      lng: Option[Double]
  ): Task[Unit]
  def deleteAddress(id: ClientAddressId, clientId: PersonId): Task[Boolean]

class ClientAddressServiceImpl(repo: ClientAddressRepository) extends ClientAddressService:

  override def getAddresses(clientId: PersonId): Task[List[ClientAddress]] = repo.findByClient(clientId)

  override def updateAddress(
      id: ClientAddressId,
      clientId: PersonId,
      req: UpdateClientAddressRequest
  ): Task[Option[ClientAddress]] = repo.updateLabelAndAliases(id, clientId, req.label, req.aliases)

  override def saveAddress(clientId: PersonId, req: SaveClientAddressRequest): Task[ClientAddress] = repo
    .findByAddressText(clientId, req.address)
    .flatMap {
      case Some(existing) => repo.incrementUseCount(existing.id).as(existing)
      case None           =>
        val addr = ClientAddress(
          id = ClientAddressId.generate(),
          clientId = clientId,
          label = req.label,
          address = req.address,
          latitude = req.latitude,
          longitude = req.longitude,
          aliases = req.aliases,
          createdAt = Instant.now(),
          updatedAt = Instant.now()
        )
        repo.save(addr)
    }

  override def recordUsage(
      clientId: PersonId,
      address: String,
      label: String,
      lat: Option[Double],
      lng: Option[Double]
  ): Task[Unit] = repo.findByAddressText(clientId, address).flatMap {
    case Some(existing) => repo.incrementUseCount(existing.id)
    case None           =>
      val addr = ClientAddress(
        id = ClientAddressId.generate(),
        clientId = clientId,
        label = label,
        address = address,
        latitude = lat,
        longitude = lng,
        createdAt = Instant.now(),
        updatedAt = Instant.now()
      )
      repo.save(addr).unit
  }

  override def deleteAddress(id: ClientAddressId, clientId: PersonId): Task[Boolean] = repo.delete(id, clientId)

object ClientAddressService:

  val layer: ZLayer[ClientAddressRepository, Nothing, ClientAddressService] = ZLayer.fromFunction(
    ClientAddressServiceImpl(_)
  )
